doc/express-reboot/Cisco_8000_Express_Reboot_HLD.md

```diff
@@ -0,0 +1,121 @@
+#
+# Sonic Express reboot HLD Spec
+
+## Overview
+
+The goal of Sonic express reboot is to be able to restart and upgrade SONiC software with sub-second data plane interruption. This document covers implementation in Sonic and SAI interface. 
+
+## fast-reboot script
+
+The existing fast-reboot script ((https://github.com/sonic-net/sonic-utilities/blob/master/scripts/fast-reboot) that is used to trigger fast-reboot/warm-boot will be enhanced to support express-reboot. Major changes are:
+
+1.	Enforce express-reboot is only applicable to Cisco 8000 platforms at the moment.
+2.	During express-reboot shutdown path, syncd_request_shutdown will be called with a new option “-pxe” to inform syncd pre-shutdown for express boot mode. The actual calls will be “docker exec syncd /usr/bin/syncd_request_shutdown –pxe”.
+3.	“SONIC_BOOT_TYPE=express” will be added to reboot cmdline to indicate express-reboot.
+
+The rest flow is similar to warm-reboot.
+
+## Syncd 
+
+Major changes in syncd components are:
+1.	A new command line option “-pxe” will be added to syncd_request_shutdown program for indication of express boot pre-shutdown. New RequestShutdownCommandLineOptions enum _syncd_restart_type_t for express boot will also be added:
+
+```
+    typedef enum _syncd_restart_type_t
+    {
+        SYNCD_RESTART_TYPE_COLD,
+        SYNCD_RESTART_TYPE_WARM,
+        SYNCD_RESTART_TYPE_FAST,
+        SYNCD_RESTART_TYPE_PRE_SHUTDOWN,
+        SYNCD_RESTART_TYPE_EXPRESS,
+        SYNCD_RESTART_TYPE_EXPRESS_PRE_SHUTDOWN,
+    } syncd_restart_type_t;
+```
+
+2.	New express boot mode will be added to syncd CommandLineOptions:
+
+```
+#define STRING_SAI_START_TYPE_EXPRESS_BOOT      "express"
+    typedef enum _sai_start_type_t
+    {
+        SAI_START_TYPE_COLD_BOOT = 0,
+        SAI_START_TYPE_WARM_BOOT = 1,
+        SAI_START_TYPE_FAST_BOOT = 2,
+
+        /**
+         * A special type of boot used by Mellanox platforms to start in 'fastfast'
+         * boot mode
+         */
+        SAI_START_TYPE_FASTFAST_BOOT = 3,
+
+        /**
+         * A special type of boot used by Cisco platforms to start in 'express'
+         * boot mode
+         */
+        SAI_START_TYPE_EXPRESS_BOOT = 4,
+
+        /**
+         * Set at last, just for error purpose.
+         */
+        SAI_START_TYPE_UNKNOWN
+    } sai_start_type_t;
+```
+
+3.	During express boot pre-shutdown, syncd sets SAI_SWITCH_ATTR_FAST_API_ENABLE = true in setPreShutdownOnAllSwitches() to notify SDK for proper cleanup.
+
+```
+sai_status_t yncd::setPreShutdownOnAllSwitches()
+{
+	    if (shutdownType == SYNCD_RESTART_TYPE_EXPRESS) 
```

Comment:
If I understand correctly, this step is the only difference between Express boot vs FastFast boot.

Why is this step needed? During Express shut, the SAI implementation must not reset the ASIC.

This can be achieved by setting SAI\_SWITCH\_ATTR\_UNINIT\_DATA\_PLANE\_ON\_REMOVAL to false.

Reply:
It needs a pre-shutdown notification so that platform hardware knows  it and does necessary things pre- and during shutdown. SAI_SWITCH_ATTR_FAST_API_ENABLE is used after boot upon receving SAI_REDIS_NOTIFY_SYNCD_APPLY_VIEW. SO we used the same attr for pre-shutdown notification as well.

Reply:
We are overloading SAI\_SWITCH\_ATTR\_FAST\_API\_ENABLE attribute. We should clearly define what the underlying SAI implementation must do for Express boot/shut.

<table><tbody><tr><td>SAI_SWITCH_ATTR_FAST_API_ENABLE</td><td>Express Boot/shut</td><td>Meaning</td></tr><tr><td>True</td><td>Boot</td><td><p>Valid only for FastFast and Express??</p><p>Signals the SAI adapter to begin the bulk programming</p><p>&nbsp;</p></td></tr><tr><td>False</td><td>Boot</td><td>Should not be used?</td></tr><tr><td>True</td><td>Shut</td><td>??</td></tr><tr><td>False</td><td>Shut</td><td>??</td></tr></tbody></table>

Reply:
Hi, this is how the SAI_SWITCH_ATTR_FAST_API_ENABLE being used for express/fastfast boot. Will add this info to the doc.

| SAI_SWITCH_ATTR_FAST_API_ENABLE | Express boot  |   fastfast boot  | Meaning 
|-----------------------------------------|-------------- |----------------|-------------
|      TRUE                                                      | Shut                 | Not used          | Pre-shutdown notification
|    FALSE                                                      | Boot                |  Boot                | Signals the SAI adapter to begin the bulk programming post reboot